internal/controller/management/serviceaccount/serviceaccounts.go

```diff
@@ -0,0 +1,289 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/logging"
+)
+
+const (
+	componentLabelKey    = "app.kubernetes.io/component"
+	controllerLabelValue = "controller"
+)
+
+type ReconcilerConfig struct {
+	KargoNamespace string `envconfig:"KARGO_NAMESPACE" required:"true"`
+}
+
+func ReconcilerConfigFromEnv() ReconcilerConfig {
+	cfg := ReconcilerConfig{}
+	envconfig.MustProcess("", &cfg)
+	return cfg
+}
+
+// reconciler reconciles ServiceAccount resources
+type reconciler struct {
+	cfg    ReconcilerConfig
+	client client.Client
+
+	getServiceAccountFn func(
+		context.Context,
+		types.NamespacedName,
+		client.Object,
+		...client.GetOption,
+	) error
+
+	createRoleBindingFn func(
+		context.Context,
+		client.Object,
+		...client.CreateOption,
+	) error
+
+	deleteRoleBindingFn func(
+		context.Context,
+		client.Object,
+		...client.DeleteOption,
+	) error
+
+	ensureControllerPermissionsFn func(context.Context, *corev1.ServiceAccount) error
+	removeControllerPermissionsFn func(context.Context, *corev1.ServiceAccount) error
+}
+
+// SetupReconcilerWithManager initializes a reconciler for ServiceAccount 
+// resources and registers it with the provided Manager.
+func SetupReconcilerWithManager(kargoMgr manager.Manager, cfg ReconcilerConfig) error {
+	return ctrl.NewControllerManagedBy(kargoMgr).
+		For(&corev1.ServiceAccount{}).
+		WithEventFilter(
+			predicate.Funcs{
+				CreateFunc: func(e event.CreateEvent) bool {
+					// Allow creation events to be handled for all ServiceAccounts
+					// This ensures any labeled or unlabeled SA gets proper
+					// reconciliation, including on restarts
+					return true
+				},
+				DeleteFunc: func(e event.DeleteEvent) bool {
+					return false
+				},
+				UpdateFunc: func(e event.UpdateEvent) bool {
+					oldSA := e.ObjectOld.(*corev1.ServiceAccount)
+					newSA := e.ObjectNew.(*corev1.ServiceAccount)
+					return hasControllerLabel(oldSA) != hasControllerLabel(newSA) || hasControllerLabel(newSA) && newSA.DeletionTimestamp != nil
+				},
+				GenericFunc: func(event.GenericEvent) bool {
+					return false
+				},
+			},
+		).
+		Complete(newReconciler(kargoMgr.GetClient(), cfg))
+}
+
+func newReconciler(kubeClient client.Client, cfg ReconcilerConfig) *reconciler {
+	r := &reconciler{
+		cfg:    cfg,
+		client: kubeClient,
+	}
+	r.getServiceAccountFn = r.client.Get
+	r.createRoleBindingFn = r.client.Create
+	r.deleteRoleBindingFn = r.client.Delete
+	r.ensureControllerPermissionsFn = r.ensureControllerPermissions
+	r.removeControllerPermissionsFn = r.removeControllerPermissions
```

Comment:
This is one of the compelling reasons to use a constructor-like function as we were previously discussing. This provides a stronger guarantee that initialization was carried out correctly.

This pattern of assigning default implementations of function fields so that they can be overridden for testing purposes is something we should discuss offline tomorrow. We may be moving away from this for _new_ components. Good for now. More to come when we chat.